src/overcooked_ai_py/visualization/extract_events.py

```diff
@@ -0,0 +1,90 @@
+import numpy as np
+import itertools, copy
+from collections import defaultdict
+from overcooked_ai_py.utils import numpy_to_native
+def extract_events(trajectories, traj_idx=0, add_cumulative_events=True):
+    """
+    extracts events from trajectories for plots
+    """
+
+    def find_last_event(events, event_type, p, last_timestep=None):
+        for event in reversed(events):
+            if event.get("player") == p and event.get("action") == event_type \
+                    and (last_timestep is None or event["timestep"] < last_timestep):
+                return event
+            
+    def holding_event(events, p, t, obj):
+        last_pickup_t = (find_last_event(events, "pickup", p, t) or {}).get("timestep")
+
+        start_t = last_pickup_t or 0
+        event = {
+            "player": p,
+            "start_timestep":start_t,
+            "end_timestep":t,
+            "action": "holding",
+            "object": obj
+        }
+        return event
+
+    def get_holding_events(events, game_states):
+        holding_events = [holding_event(events, e.get("player"), e.get("timestep"), e.get("object")) 
+            for e in events if e.get("action") in ["drop", "potting", "delivery"]]
+        # check if object is held at end of the episode
+        for p, player in enumerate(game_states[-1].players):
+            if player.held_object:
+                holding_events.append(holding_event(events, p, len(game_states)-1, player.held_object.to_dict()))
```

Comment:
Still not a fan of how this code has to "look inside" the overcooked mdp class at the game_state dictionary representation in order to function. You don't have to change it right now, because I'm hopeful that adding the trajectory class (as discussed in a different issue) will clean up this code a bunch